libcst/_metadata_dependent.py

```diff
@@ -31,6 +32,26 @@
 
 _UNDEFINED_DEFAULT = object()
 
+_SENTINEL = object()
+
+class LazyValue:
+    """
+    The class for implementing a lazy metadata loading mechanism that improves the
+    performance when retriving expensive metadata (e.g., qualified names). Providers
+    including :class:`~libcst.metadata.QualifiedNameProvider` use this class to load
+    the metadata of a certain node lazily when calling
+    :func:`~libcst.MetadataDependent.get_metadata`.
+    """
+
+    def __init__(self, callable: Callable) -> None:
```

Comment:
This will need to be a bit more strictly typed. `LazyValue` will need to be generic over some `T` value, and this callable will be typed as `Callable[[], T]`